clearnode/nitrolite/types/README.md

```diff
@@ -0,0 +1,287 @@
+
+All state transitions within the network are authorized by users and validated by a dynamic, weighted quorum of **Ledger Nodes**. A master smart contract on Ethereum maintains the network configuration, while on-chain **Custody** and **Adjudicator** contracts on each supported chain handle deposits, withdrawals, and dispute resolution.
+
+## Entities
+
+- **User**: The owner of the funds and the `State`, identified by a unique Yellow Network `UserID`. The user authorizes all actions by signing intents or states with their primary key or a delegated session key.
+- **Ledger Nodes (Clearnodes)**: A permissioned set of off-chain nodes responsible for validating user intents, issuing new states, reaching consensus, and attesting to state validity with their weighted signatures. Their registry and weights are managed by the Adjudicator contract.
+- **Master Registry Contract (on Ethereum)**: The primary on-chain contract defining the network's configuration, including approved ledger node keys, supported blockchains and the addresses of their respective Custody and Adjudicator contracts.
+- **Custody Contract (on each chain)**: Holds the pooled funds for all users on a specific blockchain, processes deposits, and executes withdrawals upon receiving valid proofs.
+- **Adjudicator Contract (on each chain)**: A companion to the Custody contract that validates state proofs. It verifies signatures, and resolves disputes by computationally verifying forced state transitions.
+
+## Architecture Components
+
+### Network Configuration
+
+The master smart contract on Ethereum serves as the central registry for network configuration:
+
+- **Registry of supported blockchains**: Defines which blockchain networks are supported
+- **Network definitions**: Maps each blockchain to its specific networks and smart contracts  
+- **Ledger node registry**: Manages approved validator nodes with signature weights
+
+```protobuf
+message NetworkConfig {
+  repeated Blockchain blockchains = 1;
+  repeated LedgerNode ledger_nodes = 2;
+}
+
+message Blockchain {
+  string name = 1;
+  repeated Network networks = 2;
+}
+
+message Network {
+  string chain_id = 1;
+  string name = 2;
+  bytes contract_address = 3;
+}
+
+message LedgerNode {
+  string node_id = 1;
+  SessionKey session_key = 2;
+  uint32 signature_weight = 3;
+}
+```
+
+### Per-Chain Configuration
+
+Each blockchain in its contract:
+
+- Maps Yellow Network asset IDs to local chain tokens
+- Maintains a registry of supported tokens
+- Mirrors the approved ledger node registry with signature weights for funds recovery
+
+```protobuf
+message ChainConfig {
+  string chain_id = 1;
+  repeated AssetToken asset_tokens = 2;
+  repeated LedgerNode ledger_nodes = 3; // Mirrors the main registry (for funds recovery)
+}
+
+message AssetToken {
+  string asset_id = 1; // Asset ID on Yellow Network
+  bytes address = 2;
+  uint32 decimals = 3;
```

Comment:
Do we need to store the decimals, if they are accessible from the token contracts?